### PR TITLE
Add regression 4014 (SM2 KEP)

### PR DESCRIPTION
**Pull request by jforissier**
Adds a test for the SM2 Key Exchange Protocol.

Signed-off-by: Jerome Forissier <jerome@forissier.org>

### CODE REVIEW

**Comment by jforissier** [comment]
Test case for https://github.com/OP-TEE/optee_os/pull/3534.

**Comment by jforissier** [comment]
Updated to include the algorithm availability test; commit "Add helper function ta_crypt_cmd_is_algo_supported()" is cherry-picked from https://github.com/OP-TEE/optee_test/pull/388.

**Comment by jforissier** [comment]
Ping!
(I fixed a silly stack corruption bug -- `params[9]` instead of `params[4]`)

File: host/xtest/regression_4000.c [hunk 1]
```diff
@@ -5543,4 +5543,335 @@ static void xtest_tee_test_4013(ADBG_Case_t *c)
...
+	TEE_ObjectHandle sv_handle = TEE_HANDLE_NULL;
+	TEE_Attribute params[9] = { };
+	size_t param_count = 0;
+	uint8_t out[128] = { };
+	size_t out_size = 0;
+	uint8_t conf_A[32];
+	uint8_t conf_B[32];
+
+	if (!ADBG_EXPECT_TEEC_SUCCESS(c,
+		xtest_teec_open_session(&session, &crypt_user_ta_uuid, NULL,
```
Comment (etienne-lms) [inline host/xtest/regression_4000.c]: prefer change indentation at lines 5565, 5582, 5589 5608 and so on.

File: host/xtest/regression_4000.c [hunk 2]
```diff
@@ -5543,4 +5543,335 @@ static void xtest_tee_test_4013(ADBG_Case_t *c)
...
+	TEE_ObjectHandle eph_keyA = TEE_HANDLE_NULL;
+	TEE_ObjectHandle keyB = TEE_HANDLE_NULL;
+	TEE_ObjectHandle eph_keyB = TEE_HANDLE_NULL;
+	TEE_ObjectHandle sv_handle = TEE_HANDLE_NULL;
+	TEE_Attribute params[9] = { };
+	size_t param_count = 0;
+	uint8_t out[128] = { };
+	size_t out_size = 0;
+	uint8_t conf_A[32];
+	uint8_t conf_B[32];
```
Comment (etienne-lms) [inline host/xtest/regression_4000.c]: maybe `= { }` for both.

File: host/xtest/regression_4000.c (hunk 1 above)
Comment (jforissier) [reply host/xtest/regression_4000.c]: With "proper" indentation (start after opening parenthesis), lines >80 characters cannot be avoided. And many tests use this style already. So I'm not sure which is best.

File: host/xtest/regression_4000.c (hunk 2 above)
Comment (jforissier) [reply host/xtest/regression_4000.c]: OK

**Comment by jforissier** [comment]
> I would have preferred some more test vector, but if that's all we have...
> `Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

There is another one in Part 3 of the spec, but no other that I know of.

File: host/xtest/regression_4000.c (hunk 1 above)
Comment (jenswi-linaro) [reply host/xtest/regression_4000.c]: We could rewrite it as:
``` C
res = xtest_teec_open_session(...)
if (!ADBG_EXPECT_TEEC_SUCCESS(c, res))
         return;
```
However, I'm not sure that should be in the scope of this PR.

Comment (jforissier) [reply host/xtest/regression_4000.c]: @jenswi-linaro true, we have about 800 occurrences of this pattern... all are more or less formatted like the above.

**Comment by jforissier** [comment]
Update.

**Pull request merged**